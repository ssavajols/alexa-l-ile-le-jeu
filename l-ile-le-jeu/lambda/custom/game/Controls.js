"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Controls = /** @class */ (function () {
    function Controls(handlerInput, _actionType) {
        this._actionType = _actionType || '';
        try {
            this._slots = handlerInput.requestEnvelope.request.intent.slots;
        }
        catch (err) {
            this._slots = {};
        }
    }
    Controls.prototype.getAction = function () {
        var actionType = this._actionType;
        var actionName = this._slotValue('action');
        if (actionName && actionType) {
            return actionType.toUpperCase() + "_" + actionName.toUpperCase();
        }
        else if (actionName) {
            return "" + actionName.toUpperCase();
        }
        return null;
    };
    Controls.prototype._slotValue = function (slotName) {
        return this._slots && this._slots[slotName] ? this._slots[slotName].value : null;
    };
    return Controls;
}());
exports.Controls = Controls;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiQ29udHJvbHMuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJDb250cm9scy50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUVBO0lBTUUsa0JBQWEsWUFBZ0MsRUFBRSxXQUFvQjtRQUVqRSxJQUFJLENBQUMsV0FBVyxHQUFHLFdBQVcsSUFBSSxFQUFFLENBQUE7UUFFcEMsSUFBSTtZQUNGLElBQUksQ0FBQyxNQUFNLEdBQUksWUFBWSxDQUFDLGVBQWUsQ0FBQyxPQUFlLENBQUMsTUFBTSxDQUFDLEtBQUssQ0FBQTtTQUN6RTtRQUFDLE9BQU8sR0FBRyxFQUFFO1lBQ1osSUFBSSxDQUFDLE1BQU0sR0FBRyxFQUFFLENBQUE7U0FDakI7SUFDSCxDQUFDO0lBRU0sNEJBQVMsR0FBaEI7UUFDRSxJQUFNLFVBQVUsR0FBRyxJQUFJLENBQUMsV0FBVyxDQUFBO1FBQ25DLElBQU0sVUFBVSxHQUFHLElBQUksQ0FBQyxVQUFVLENBQUMsUUFBUSxDQUFDLENBQUE7UUFFNUMsSUFBSSxVQUFVLElBQUksVUFBVSxFQUFFO1lBQzVCLE9BQVUsVUFBVSxDQUFDLFdBQVcsRUFBRSxTQUFJLFVBQVUsQ0FBQyxXQUFXLEVBQUksQ0FBQTtTQUNqRTthQUFNLElBQUksVUFBVSxFQUFFO1lBQ3JCLE9BQU8sS0FBRyxVQUFVLENBQUMsV0FBVyxFQUFJLENBQUE7U0FDckM7UUFFRCxPQUFPLElBQUksQ0FBQTtJQUNiLENBQUM7SUFFTyw2QkFBVSxHQUFsQixVQUFvQixRQUFnQjtRQUNsQyxPQUFPLElBQUksQ0FBQyxNQUFNLElBQUksSUFBSSxDQUFDLE1BQU0sQ0FBQyxRQUFRLENBQUMsQ0FBQyxDQUFDLENBQUMsSUFBSSxDQUFDLE1BQU0sQ0FBQyxRQUFRLENBQUMsQ0FBQyxLQUFLLENBQUMsQ0FBQyxDQUFDLElBQUksQ0FBQTtJQUNsRixDQUFDO0lBQ0gsZUFBQztBQUFELENBQUMsQUFqQ0QsSUFpQ0M7QUFqQ1ksNEJBQVEifQ==