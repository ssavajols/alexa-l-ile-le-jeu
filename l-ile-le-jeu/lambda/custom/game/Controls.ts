import * as Alexa from 'ask-sdk-core'

export class Controls {

  private _slots: any

  private _actionType: string

  constructor (handlerInput: Alexa.HandlerInput, _actionType?: string) {

    this._actionType = _actionType || ''

    try {
      this._slots = (handlerInput.requestEnvelope.request as any).intent.slots
    } catch (err) {
      this._slots = {}
    }
  }

  public getAction (): string | null {
    const actionType = this._actionType
    const actionName = this._slotValue('action')

    if (actionName && actionType) {
      return `${actionType.toUpperCase()}_${actionName.toUpperCase()}`
    } else if (actionName) {
      return `${actionName.toUpperCase()}`
    }

    return null
  }

  private _slotValue (slotName: string): any {
    return this._slots && this._slots[slotName] ? this._slots[slotName].value : null
  }
}
