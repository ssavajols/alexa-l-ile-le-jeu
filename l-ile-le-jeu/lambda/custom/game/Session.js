"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Session = /** @class */ (function () {
    function Session(handlerInput) {
        this._session = handlerInput.attributesManager;
    }
    Session.prototype.getAttribute = function (name) {
        return this._session.getSessionAttributes()[name];
    };
    Session.prototype.setAttribute = function (name, value) {
        var attr = this._session.getSessionAttributes();
        attr[name] = value;
        this._session.setSessionAttributes(attr);
    };
    return Session;
}());
exports.Session = Session;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiU2Vzc2lvbi5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIlNlc3Npb24udHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6Ijs7QUFFQTtJQUlFLGlCQUFhLFlBQWdDO1FBQzNDLElBQUksQ0FBQyxRQUFRLEdBQUcsWUFBWSxDQUFDLGlCQUFpQixDQUFBO0lBQ2hELENBQUM7SUFFTSw4QkFBWSxHQUFuQixVQUFxQixJQUFZO1FBQy9CLE9BQU8sSUFBSSxDQUFDLFFBQVEsQ0FBQyxvQkFBb0IsRUFBRSxDQUFDLElBQUksQ0FBQyxDQUFBO0lBQ25ELENBQUM7SUFFTSw4QkFBWSxHQUFuQixVQUFxQixJQUFZLEVBQUUsS0FBVTtRQUMzQyxJQUFNLElBQUksR0FBRyxJQUFJLENBQUMsUUFBUSxDQUFDLG9CQUFvQixFQUFFLENBQUE7UUFFakQsSUFBSSxDQUFDLElBQUksQ0FBQyxHQUFHLEtBQUssQ0FBQTtRQUVsQixJQUFJLENBQUMsUUFBUSxDQUFDLG9CQUFvQixDQUFDLElBQUksQ0FBQyxDQUFBO0lBQzFDLENBQUM7SUFDSCxjQUFDO0FBQUQsQ0FBQyxBQW5CRCxJQW1CQztBQW5CWSwwQkFBTyJ9