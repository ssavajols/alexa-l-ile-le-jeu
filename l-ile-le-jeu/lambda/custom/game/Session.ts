import * as Alexa from 'ask-sdk-core'

export class Session {

  private _session: Alexa.AttributesManager

  constructor (handlerInput: Alexa.HandlerInput) {
    this._session = handlerInput.attributesManager
  }

  public getAttribute (name: string): any {
    return this._session.getSessionAttributes()[name]
  }

  public setAttribute (name: string, value: any) {
    const attr = this._session.getSessionAttributes()

    attr[name] = value

    this._session.setSessionAttributes(attr)
  }

  public getNewInt (): number {
    const int = typeof this.getAttribute('int') === 'number' ? this.getAttribute('int') + 1 : 1
    this.setAttribute('int', int)
    return int
  }
}
