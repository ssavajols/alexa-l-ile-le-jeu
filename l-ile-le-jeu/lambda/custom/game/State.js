"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var errors_1 = require("./errors");
var QuestionEntity_1 = require("./QuestionEntity");
var State = /** @class */ (function () {
    function State(data, progress) {
        this._progress = '';
        this._data = data;
        this.setProgress(progress);
    }
    State.prototype.next = function (actionName) {
        if (!this.getProgress()) {
            this.setProgress('Q1');
            return this.getIntroductionSpeech();
        }
        else if (actionName) {
            this.doAction(actionName);
            return this.getQuestionSpeech();
        }
        else {
            return this.getQuestionSpeech();
        }
    };
    State.prototype.getProgress = function () {
        return this._progress;
    };
    State.prototype.setProgress = function (progress) {
        if (progress) {
            this._progress = progress;
            this.setQuestion();
        }
        return this.getProgress();
    };
    State.prototype.updateProgress = function (progress) {
        if (progress) {
            this._progress += this._progress.length ? "." + progress : progress;
            this.setQuestion();
        }
        return this.getProgress();
    };
    State.prototype.getQuestion = function () {
        return this._question;
    };
    State.prototype.doAction = function (actionName) {
        if (!this._question) {
            throw new errors_1.QuestionNotDefined();
        }
        this.setProgress("" + this.getProgress() + this.getActionIndex(actionName));
        return this._question;
    };
    State.prototype.getActionIndex = function (actionName) {
        if (!this._question) {
            throw new errors_1.QuestionNotDefined();
        }
        var answerIndex = this._question.findAnswerIndexByValue(actionName) + 1;
        var currentQuestion = this.getProgress().split('.').pop();
        return answerIndex > 0 ? "." + (currentQuestion + answerIndex.toString()) : '';
    };
    State.prototype.getIntroductionSpeech = function () {
        return "<p>" + this._data.introduction + "</p>" + this.getQuestionSpeech();
    };
    State.prototype.getQuestionSpeech = function () {
        return this._question ? this._question.getQuestionSpeech() : '';
    };
    State.prototype.setQuestion = function () {
        return this._question = new QuestionEntity_1.QuestionEntity(this._data, this._progress.split('.').pop());
    };
    return State;
}());
exports.State = State;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiU3RhdGUuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJTdGF0ZS50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUFBLG1DQUE2QztBQUM3QyxtREFBaUQ7QUFFakQ7SUFRRSxlQUFhLElBQUksRUFBRSxRQUFTO1FBSnBCLGNBQVMsR0FBVyxFQUFFLENBQUE7UUFLNUIsSUFBSSxDQUFDLEtBQUssR0FBRyxJQUFJLENBQUE7UUFDakIsSUFBSSxDQUFDLFdBQVcsQ0FBQyxRQUFRLENBQUMsQ0FBQTtJQUM1QixDQUFDO0lBRU0sb0JBQUksR0FBWCxVQUFhLFVBQW1CO1FBQzlCLElBQUksQ0FBQyxJQUFJLENBQUMsV0FBVyxFQUFFLEVBQUU7WUFDdkIsSUFBSSxDQUFDLFdBQVcsQ0FBQyxJQUFJLENBQUMsQ0FBQTtZQUN0QixPQUFPLElBQUksQ0FBQyxxQkFBcUIsRUFBRSxDQUFBO1NBQ3BDO2FBQU0sSUFBSSxVQUFVLEVBQUU7WUFDckIsSUFBSSxDQUFDLFFBQVEsQ0FBQyxVQUFVLENBQUMsQ0FBQTtZQUN6QixPQUFPLElBQUksQ0FBQyxpQkFBaUIsRUFBRSxDQUFBO1NBQ2hDO2FBQU07WUFDTCxPQUFPLElBQUksQ0FBQyxpQkFBaUIsRUFBRSxDQUFBO1NBQ2hDO0lBQ0gsQ0FBQztJQUVNLDJCQUFXLEdBQWxCO1FBQ0UsT0FBTyxJQUFJLENBQUMsU0FBUyxDQUFBO0lBQ3ZCLENBQUM7SUFFTSwyQkFBVyxHQUFsQixVQUFvQixRQUFnQjtRQUNsQyxJQUFJLFFBQVEsRUFBRTtZQUNaLElBQUksQ0FBQyxTQUFTLEdBQUcsUUFBUSxDQUFBO1lBQ3pCLElBQUksQ0FBQyxXQUFXLEVBQUUsQ0FBQTtTQUNuQjtRQUVELE9BQU8sSUFBSSxDQUFDLFdBQVcsRUFBRSxDQUFBO0lBQzNCLENBQUM7SUFFTSw4QkFBYyxHQUFyQixVQUF1QixRQUFnQjtRQUNyQyxJQUFJLFFBQVEsRUFBRTtZQUNaLElBQUksQ0FBQyxTQUFTLElBQUksSUFBSSxDQUFDLFNBQVMsQ0FBQyxNQUFNLENBQUMsQ0FBQyxDQUFDLE1BQUksUUFBVSxDQUFDLENBQUMsQ0FBQyxRQUFRLENBQUE7WUFDbkUsSUFBSSxDQUFDLFdBQVcsRUFBRSxDQUFBO1NBQ25CO1FBRUQsT0FBTyxJQUFJLENBQUMsV0FBVyxFQUFFLENBQUE7SUFDM0IsQ0FBQztJQUVNLDJCQUFXLEdBQWxCO1FBQ0UsT0FBTyxJQUFJLENBQUMsU0FBUyxDQUFBO0lBQ3ZCLENBQUM7SUFFTSx3QkFBUSxHQUFmLFVBQWlCLFVBQWtCO1FBQ2pDLElBQUksQ0FBQyxJQUFJLENBQUMsU0FBUyxFQUFFO1lBQ25CLE1BQU0sSUFBSSwyQkFBa0IsRUFBRSxDQUFBO1NBQy9CO1FBQ0QsSUFBSSxDQUFDLFdBQVcsQ0FBQyxLQUFHLElBQUksQ0FBQyxXQUFXLEVBQUUsR0FBRyxJQUFJLENBQUMsY0FBYyxDQUFDLFVBQVUsQ0FBRyxDQUFDLENBQUE7UUFDM0UsT0FBTyxJQUFJLENBQUMsU0FBUyxDQUFBO0lBQ3ZCLENBQUM7SUFFTyw4QkFBYyxHQUF0QixVQUF3QixVQUFrQjtRQUN4QyxJQUFJLENBQUMsSUFBSSxDQUFDLFNBQVMsRUFBRTtZQUNuQixNQUFNLElBQUksMkJBQWtCLEVBQUUsQ0FBQTtTQUMvQjtRQUNELElBQU0sV0FBVyxHQUFHLElBQUksQ0FBQyxTQUFTLENBQUMsc0JBQXNCLENBQUMsVUFBVSxDQUFDLEdBQUcsQ0FBQyxDQUFBO1FBQ3pFLElBQU0sZUFBZSxHQUFHLElBQUksQ0FBQyxXQUFXLEVBQUUsQ0FBQyxLQUFLLENBQUMsR0FBRyxDQUFDLENBQUMsR0FBRyxFQUFFLENBQUE7UUFDM0QsT0FBTyxXQUFXLEdBQUcsQ0FBQyxDQUFDLENBQUMsQ0FBQyxPQUFJLGVBQWUsR0FBRyxXQUFXLENBQUMsUUFBUSxFQUFFLENBQUUsQ0FBQyxDQUFDLENBQUMsRUFBRSxDQUFBO0lBQzlFLENBQUM7SUFFTyxxQ0FBcUIsR0FBN0I7UUFDRSxPQUFPLFFBQU0sSUFBSSxDQUFDLEtBQUssQ0FBQyxZQUFZLFlBQU8sSUFBSSxDQUFDLGlCQUFpQixFQUFJLENBQUE7SUFDdkUsQ0FBQztJQUVPLGlDQUFpQixHQUF6QjtRQUNFLE9BQU8sSUFBSSxDQUFDLFNBQVMsQ0FBQyxDQUFDLENBQUMsSUFBSSxDQUFDLFNBQVMsQ0FBQyxpQkFBaUIsRUFBRSxDQUFDLENBQUMsQ0FBQyxFQUFFLENBQUE7SUFDakUsQ0FBQztJQUVPLDJCQUFXLEdBQW5CO1FBQ0UsT0FBTyxJQUFJLENBQUMsU0FBUyxHQUFHLElBQUksK0JBQWMsQ0FBQyxJQUFJLENBQUMsS0FBSyxFQUFFLElBQUksQ0FBQyxTQUFTLENBQUMsS0FBSyxDQUFDLEdBQUcsQ0FBQyxDQUFDLEdBQUcsRUFBRSxDQUFDLENBQUE7SUFDekYsQ0FBQztJQUVILFlBQUM7QUFBRCxDQUFDLEFBaEZELElBZ0ZDO0FBaEZZLHNCQUFLIn0=