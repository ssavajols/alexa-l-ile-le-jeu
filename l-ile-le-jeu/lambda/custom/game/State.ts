// const { log } = console

export class State {

  private _data: any

  private _progress: string = ''

  constructor (data) {
    this._data = data
  }

  public next (actionName?: string) {
    if (!this.getProgress()) {
      this.setProgress('Q1')
      return this.getIntroductionSpeech()
    } else if (actionName) {
      return this.getQuestionSpeech(this.doAction(actionName))
    } else {
      return this.getQuestionSpeech()
    }
  }

  public getProgress () {
    return this._progress
  }

  public setProgress (progress: string): string {
    if (progress) {
      this._progress = progress
    }

    return this.getProgress()
  }

  public updateProgress (progress: string): string {
    if (progress) {
      this._progress += this._progress.length ? `.${progress}` : progress
    }

    return this.getProgress()
  }

  public doAction (actionName) {
    this.setProgress(`${this.getProgress()}${this.getActionIndex(actionName)}`)
    return this.getQuestion(this.getProgress())
  }

  private getActionIndex (actionName) {
    const question = this.getQuestion()
    const progress = this.getProgress()
    const aProgress = progress.split('.')
    const answerIndex = question.answers.indexOf(actionName) + 1
    const currentQuestion = aProgress[aProgress.length - 1]
    return answerIndex > 0 ? `.${currentQuestion + answerIndex.toString()}` : ''
  }

  private getIntroductionSpeech () {
    return `<p>${this._data.introduction}</p>${this.getQuestionSpeech()}`
  }

  private getQuestionSpeech (question = this.getQuestion()) {
    return `<p>${question.question}</p><p>${question.answers.join('\r')}</p>`
  }

  private getQuestion (progress = this.getProgress()) {
    return progress.split('.').reduce((p, c) => p[c], this._data)
  }

}
