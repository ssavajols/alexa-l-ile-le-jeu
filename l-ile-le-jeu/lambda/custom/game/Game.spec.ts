import { Game } from '.'
import { HandlerInputFactory } from '../../../../__tests__/HandlerInputFactory'
import { Controls } from './Controls'
import * as data from './data/state-fr-FR.spec.json'
import { Session } from './Session'
import { State } from './State'

describe('GAME', () => {

  let GAME
  let handlerInput

  beforeEach(() => {

    handlerInput = HandlerInputFactory.create({
      action: { value: 'Droite' }
    }, {
      progress: 'Q1.Q12'
    })
    GAME = new Game(handlerInput, 'Take')
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should contain session instance', () => {
    expect(GAME.session instanceof Session).toBe(true)
  })

  it('should contain controls instance', () => {
    expect(GAME.controls instanceof Controls).toBe(true)
  })

  it('should contain state instance', () => {
    expect(GAME.state instanceof State).toBe(true)
  })

  it('should return help message', () => {
    expect(GAME.help()).toBe('help message must be here !')
  })

  it('should return skill name', () => {
    expect(GAME.skillName()).toBe('le jeu')
  })

  it('should return stop message', () => {
    expect(GAME.stop()).toBe('Au revoir !')
  })

  it('should return fallback message', () => {
    expect(GAME.fallback()).toBe('Commande non reconnu')
  })

  it('should get Q12', () => {
    expect(GAME.getSpeech()).toMatchSnapshot()
  })

  it('should get isEnd attribute => false', () => {
    expect(GAME.isEnd).toBeFalsy()
  })

  it('should get isEnd attribute => true', () => {
    const h = HandlerInputFactory.create({}, {
      progress: 'Q2'
    })
    const game = new Game(h)
    expect(game.isEnd).toBeTruthy()
    expect(game.getSpeech()).toMatchSnapshot()
  })

})
