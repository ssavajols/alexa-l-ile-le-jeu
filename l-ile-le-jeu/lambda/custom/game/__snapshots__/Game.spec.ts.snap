// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GAME should get Q12 1`] = `"<p>tata ?</p><p>TAKE_CHAISE. MOVE_BAS</p>"`;

exports[`GAME should get isEnd attribute => true 1`] = `"<p>No way out ?</p>"`;
